dask/tests/test_tokenize.py

```diff
@@ -310,42 +346,173 @@ def test_normalize_base():
 
 
 def test_tokenize_object():
-    check_tokenize(object(), idempotent=False)
+    with dask.config.set({"tokenize.ensure-deterministic": False}):
+        # object() tokenization is idempotent...
+        o = object()
+        assert tokenize(o) == tokenize(o)
+        # ...but not deterministic
+        assert tokenize(object()) != tokenize(object())
+
+        # Two objects don't tokenize to the same token even if their pickle output is
+        # identical. Stress id reuse by creating and dereferencing many objects in quick
+        # succession.
+        assert len({tokenize(object()) for _ in range(100)}) == 100
+
+        # You can call normalize_token even if the _ensure_deterministic context
+        # variable hasn't been set
+        assert normalize_token(o) == normalize_token(o)
+
+    with dask.config.set({"tokenize.ensure-deterministic": True}):
+        with pytest.raises(TokenizationError, match="deterministic"):
+            tokenize(o)
+        with pytest.raises(TokenizationError, match="deterministic"):
+            normalize_token(o)
+
+    # Test env override
+    assert tokenize(o, ensure_deterministic=False) == tokenize(
+        o, ensure_deterministic=False
+    )
+    with pytest.raises(TokenizationError, match="deterministic"):
+        tokenize(o, ensure_deterministic=True)
+
+    # Test ctx variable cleanup
+    with pytest.raises(LookupError):
+        _ensure_deterministic.get()
+
+
+def nested_tokenize_ensure_deterministic():
+    """Test that the ensure_deterministic override is not lost if tokenize() is
+    called recursively
+    """
+
+    class C:
+        def __dask_tokenize__(self):
+            return tokenize(object())
+
+    assert tokenize(C(), ensure_deterministic=False) != tokenize(
+        C(), ensure_deterministic=False
+    )
+    with pytest.raises(TokenizationError):
+        tokenize(C())
+
+    # Test ctx variable cleanup
+    with pytest.raises(LookupError):
+        _ensure_deterministic.get()
 
 
 _GLOBAL = 1
 
 
-def test_tokenize_local_functions():
-    a, b, c, d, e = (
-        # Note: Same line, same code lambdas cannot be distinguished
+def _local_functions():
+    all_funcs = [
         lambda x: x,
         lambda x: x + 1,
-        lambda x: x,
```

Comment:
why remove that?

Reply:
This PR incorporates https://github.com/dask/dask/pull/10883.
To review this PR before the above is merged, just look at the last commit.
This comment is about a change in #10883; please keep the critique there.